src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -2596,5 +2631,375 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+            {
+                Etag = attachmenEtag,
+                ContentType = contentType,
+                Name = name,
+                DocumentId = documentId, // TODO: Return the documentId from the storage, which may be with a different casing.
+            };
+        }
+
+        private bool UpdateDocumentAfterAttachmentWrite(DocumentsOperationContext context, byte* lowerDocumentId, int lowerDocumentIdSize, string documentId, string name, 
+            long newEtag, long modifiedTicks, out CollectionName collectionName, bool isDelete, long? expectedEtag)
+        {
+            Slice loweredKey;
+            using (Slice.External(context.Allocator, lowerDocumentId, lowerDocumentIdSize, out loweredKey))
+            {
+                var readTable = new Table(DocsSchema, context.Transaction.InnerTransaction);
+                TableValueReader tvr;
+                if (readTable.ReadByKey(loweredKey, out tvr) == false)
+                {
+                    collectionName = null;
+                    return ThrowDocumentNotFound(context, loweredKey, documentId, name, isDelete, expectedEtag);
+                }
+
+                int size;
+                var data = new BlittableJsonReaderObject(tvr.Read((int) DocumentsTable.Data, out size), size, context);
+                collectionName = ExtractCollectionName(context, documentId, data);
+                var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collectionName.GetTableName(CollectionTableType.Documents));
+                if (table.ReadByKey(loweredKey, out tvr) == false)
+                {
+                    return ThrowDocumentNotFound(context, loweredKey, documentId, name, isDelete, expectedEtag);
+                }
+
+                var newEtagBigEndian = Bits.SwapBytes(newEtag);
+
+                var oldChangeVector = GetChangeVectorEntriesFromTableValueReader(ref tvr, 4);
+                var changeVector = SetDocumentChangeVectorForLocalChange(context, loweredKey, oldChangeVector, newEtag);
+
+                fixed (ChangeVectorEntry* pChangeVector = changeVector)
+                {
+                    var transactionMarker = context.GetTransactionMarker();
+                    var flags = *(DocumentFlags*) tvr.Read((int) DocumentsTable.Flags, out size) | DocumentFlags.HasAttachments;
+                    var tbv = new TableValueBuilder
+                    {
+                        {tvr.Read((int) DocumentsTable.LoweredKey, out size), size},
+                        newEtagBigEndian,
+                        {tvr.Read((int) DocumentsTable.Key, out size), size},
+                        {data.BasePointer, data.Size},
+                        {(byte*) pChangeVector, sizeof(ChangeVectorEntry) * changeVector.Length},
+                        modifiedTicks,
+                        (int) flags,
+                        transactionMarker
+                    };
+
+                    table.Update(tvr.Id, tbv);
```

Comment:
In this case I think it safe, because every call of this function "data" and "LoweredKey" are a new objects (btw, do you need to dispose the data object?).
So we can safely update the table without corruption.